lib/smooth/include/smooth/core/io/UART.h

```diff
@@ -0,0 +1,424 @@
+    /// \param wakeup_threshold  number of RX edges for light sleep wakeup, value is 3 .. 0x3ff.
+    bool set_wakeup_threshold(int threshold);
+
+    /// Wait until UART tx memory empty and the last char send ok (polling mode).
+    /// \return true on success, false on failure
+    bool wait_tx_idle_polling();
+
+    /// Configure TX signal loop back to RX module, just for the test usage.
+    /// \param enable Set true to enable the loop back function, else set it false.
+    /// \return true on success, false on failure
+    bool set_loop_back(bool enable);
+
+    /// Configure behavior of UART RX timeout interrupt.
+    ///
+    /// When always_rx_timeout is true, timeout interrupt is triggered even if FIFO is full.
+    /// This function can cause extra timeout interrupts triggered only to send the timeout event.
+    /// Call this function only if you want to ensure timeout interrupt will always happen after a byte stream.
+    /// \param enable Set to false enable the default behavior of timeout interrupt,
+    ///               set it to true to always trigger timeout interrupt.
+    /// \return true on success, false on failure
+    void set_always_rx_timeout(bool enable);
+
+  private:
+    uart_port_t port_;
+
+    uart_config_t config_;
+    
+    /// Install UART driver and set the UART to the default configuration.
+    ///
+    /// UART ISR handler will be attached to the same CPU core that this function is running on.
+    ///
+    /// \note Rx_buffer_size should be greater than UART_FIFO_LEN. Tx_buffer_size should be either zero or greater than UART_FIFO_LEN.
+    /// 
+    /// \param rx_buffer_size UART RX ring buffer size.
+    /// \param tx_buffer_size UART TX ring buffer size.
+    ///        If set to zero, driver will not use TX buffer, TX function will block task until all data have been sent out.
+    /// \param queue_size UART event queue size/depth.
+    /// \param uart_queue UART event queue handle (out param). On success, a new queue handle is written here to provide
+    ///        access to UART events. If set to NULL, driver will not use an event queue.
+    /// \param intr_alloc_flags Flags used to allocate the interrupt. One or multiple (ORred)
+    ///        ESP_INTR_FLAG_* values. See esp_intr_alloc.h for more info. Do not set ESP_INTR_FLAG_IRAM here
+    ///        (the driver's ISR handler is not located in IRAM)
+    bool driver_install(int rx_buffer_size, int tx_buffer_size, int queue_size, QueueHandle_t* uart_queue, int intr_alloc_flags);
+    
+    /// Uninstall UART driver.
+    /// \return true on success, false on failure
+    bool driver_delete();
+    
+    /// Get the UART driver installed flag.
+    /// \return true if the UART driver is installed, false if not.
+    bool is_driver_installed() const;
+
+    /// Log an error message.
+    /// \param msg The error message
+    /// \param err The error code
+    void log_error(const std::string msg, esp_err_t err);
```

Comment:
```suggestion
    void log_error(const std::string& msg, esp_err_t err);
```
No reason to pass this function a copy of the string, no use of move-semantics that I can see?